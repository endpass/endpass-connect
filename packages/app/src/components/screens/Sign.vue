<template>
  <SignForm 
    :accounts="accounts"
    :loading="loading"
    :request="request"
    @cancel="handleSignCancel"
    @submit="handleSignSubmit"
  />
</template>

<script>
import { mapActions, mapMutations, mapState } from 'vuex'
import SignForm from '@/components/SignForm.vue';

export default {
  name: 'SignScreen',

  computed: {
    ...mapState({
      accounts: state => state.accounts.accounts,
      inited: state => state.core.inited,
      loading: state => state.core.loading,
      request: state => state.requests.request
    }),
  },

  methods: {
    ...mapActions(['awaitRequestMessage', 'sendMessage', 'closeDialog', 'processRequest']),

    sendMessageAndClose(message) {
      this.sendMessage(message)
      this.closeDialog()
    },

    async handleSignSubmit(res) {
      try {
        await this.processRequest(res.password)
      } catch (err) {
        console.log('sign failed: ', err)
      }
    },

    handleSignCancel() {
      this.sendMessageAndClose({
        message: 'User cancel sign',
        status: false,  
      })
    }
  },

  async created() {
    try {
      // Send ready message to opener for determine application ready
      await this.sendMessage({
        method: 'connect_ready',
        status: true
      })
      this.awaitRequestMessage()
    } catch (err) {
      console.log('Sign created error: ', err)
    }
  },

  components: {
    SignForm
  }
}
</script>