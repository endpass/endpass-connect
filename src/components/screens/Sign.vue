<template>
  <v-frame :loading="!request">    
    <sign-form 
      :accounts="accounts"
      :loading="loading"
      :request="request"
      :error="error"
      :closable="isDialog"
      @cancel="handleSignCancel"
      @submit="handleSignSubmit"
    />
  </v-frame> 
</template>

<script>
import { mapActions, mapGetters, mapState } from 'vuex';
import VFrame from '../VFrame.vue';
import SignForm from '../SignForm.vue';

export default {
  name: 'Sign',

  data: () => ({
    error: null,
  }),

  computed: {
    ...mapState({
      accounts: state => state.accounts.accounts,
      inited: state => state.core.inited,
      loading: state => state.core.loading,
      request: state => state.requests.request,
    }),
    ...mapGetters(['isDialog']),
  },

  methods: {
    ...mapActions(['awaitRequestMessage', 'processRequest', 'cancelRequest']),

    async handleSignSubmit(res) {
      try {
        await this.processRequest(res.password);
        this.error = null;
      } catch (err) {
        this.error = err.message;
      }
    },

    handleSignCancel() {
      this.cancelRequest();
    },

    handleWindowClose() {
      this.cancelRequest();
    },
  },

  async created() {
    if (this.isDialog) {
      window.addEventListener('beforeunload', this.handleWindowClose);
      this.awaitRequestMessage();
    }
  },

  components: {
    SignForm,
    VFrame,
  },
};
</script>
