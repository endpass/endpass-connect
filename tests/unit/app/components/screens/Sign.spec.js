import Vuex from 'vuex';
import { shallowMount, createLocalVue } from '@vue/test-utils';
import Sign from '@@/app/src/components/screens/Sign.vue';

const localVue = createLocalVue();

localVue.use(Vuex);

describe('Sign', () => {
  describe('render', () => {
    let store;
    let storeData;
    let wrapper;
    let accountsModule;
    let requestsModule;
    let coreModule;

    beforeEach(() => {
      coreModule = {
        state: {
          inited: true,
          loading: false,
        },
        actions: {
          sendReadyMessage: jest.fn(),
        },
      };
      accountsModule = {
        state: {
          accounts: [],
        },
      };
      requestsModule = {
        state: {
          request: {},
        },
        actions: {
          awaitRequestMessage: jest.fn(),
          processRequest: jest.fn(),
          cancelRequest: jest.fn(),
        },
      };
      storeData = {
        modules: {
          accounts: accountsModule,
          core: coreModule,
          requests: requestsModule,
        },
      };
      store = new Vuex.Store(storeData);
      wrapper = shallowMount(Sign, {
        localVue,
        store,
      });
    });

    describe('render', () => {
      it('should correctly render Sign component', () => {
        expect(wrapper.name()).toBe('Sign');
        expect(wrapper.html()).toMatchSnapshot();
      });
    });

    describe('behavior', () => {
      it('should confirm current request with password', () => {
        // TODO Have troubles with triggering event from stub, solve it when possivble
        wrapper.vm.handleSignSubmit({
          password: 'foo',
        });

        expect(requestsModule.actions.processRequest).toBeCalled();
      });

      it('should cancel current request', () => {
        // TODO Have troubles with triggering event from stub, solve it when possivble
        wrapper.vm.handleSignCancel();

        expect(requestsModule.actions.cancelRequest).toBeCalled();
      });
    });
  });
});
