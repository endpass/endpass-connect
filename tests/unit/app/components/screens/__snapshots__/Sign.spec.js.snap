// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sign render render should correctly render Sign component 1`] = `
<v-frame-stub>
  <sign-form-stub request="[object Object]" accounts="" closable="true"></sign-form-stub>
</v-frame-stub>
`;
